content/doc/book/pipeline/pipeline-best-practices.adoc

```diff
@@ -0,0 +1,95 @@
+Lastly, because of the ubiquitous use of these steps throughout Pipelines, if something is coded incorrectly/inefficiently the results can be catastrophic to Jenkins.
+
+=== Avoiding large global variable declaration files
+
+Having large variable declaration files can require large amounts of memory for little to no benefit, because the file is loaded for every Pipeline whether the variables are needed or not. Creating small variable files that contain only variables relevant to the current execution is recommended.
+It would be better to load variables which are relevant to the current execution.
+
+=== Avoiding very large shared libraries
+
+Using large shared libraries in Pipelines requires checking out a very large file before the Pipeline can start and loading the same shared library per job that is currently executing, which can lead to increased memory overhead and slower execution time.
+
+== Answering additional FAQs
+
+=== Dealing with Concurrency in Pipelines
+
+Try not to share workspaces across multiple Pipeline executions or multiple distinct Pipelines. 
+This practice can lead to either unexpected file modification within each Pipeline or workspace renaming.
+
+Ideally, shared volumes/disks are mounted in a separate place and the files are copied from that place to the current workspace.
+Then when the build is done the files can be copied back if there were updates done.
+
+Build in distinct containers which create needed resources from scratch(cloud-type agents work great for this). 
+Building these containers will ensure that the build process begins at the start every time and is easily repeatable.
+If building containers will not work, disable concurrency on the Pipeline or use the Lockable Resources Plugin to lock the workspace when it is running so that no other builds can use it while it is locked.
+**WARNING**: Disabling concurrency or locking the workspace when it is running can cause Pipelines to become blocked when waiting on resources if those resources are arbitrarily locked.
+
+**Also be aware that both of these slow down the time to results of builds over being able to use unique resources for each job**
+
+=== Using @NonCPS
+
+Pipeline code is CPS-transformed so that Pipelines are able to resume after a Jenkins restart. Some Groovy expressions do not work correctly as a result of this transformation. See http://jenkins.io/redirect/pipeline-cps-method-mismatches for more details and some examples of things that may be problematic. If necessary, you can use the @NonCPS annotation to disable the CPS transformation for a specific method whose body would not execute correctly if it were CPS-transformed. Just be aware that this also means the Groovy function will have to restart completely since it is not transformed.
+
+**Note that asynchronous Pipeline steps (such as sh and sleep) are always CPS-transformed, and may not be used inside of a method annotated with @NonCPS.**
```

Comment:
**NOTE:** Asynchronous Pipeline steps (such as `sh` and `sleep`) are always CPS-transformed and may not be used inside of a method annotated with @NonCPS.